File: apps/cli/src/commands/interact.rs
use crate::cli::{Cli, InvokeArgs, SendArgs, StateArgs};
use crate::commands::{cache, deploy, detect, target};
use crate::error::{AppError, AppResult};
use crate::output::{self, Meta};
use serde::Serialize;
use serde_json::Value;
use std::fs;
use std::process::Command;

#[derive(Debug, Serialize)]
struct CallData {
    contract: String,
    address: String,
    function: String,
    signature: String,
    raw: String,
}

#[derive(Debug, Serialize)]
struct SendData {
    contract: String,
    address: String,
    function: String,
    signature: String,
    tx_output: String,
    gas_estimate: Option<String>,
}

#[derive(Debug, Serialize)]
struct StateData {
    contract: String,
    address: String,
    values: Vec<StateValue>,
}

#[derive(Debug, Serialize)]
struct StateValue {
    name: String,
    signature: String,
    raw: String,
}

pub fn call(cli: &Cli, args: &InvokeArgs) -> AppResult<()> {
    let context = context(cli, &args.target)?;
    let signature = resolve_function_signature(&context.artifact, &args.function, false)?;
    let raw = cast_call(
        &context.address,
        &signature,
        &args.args,
        &context.network.rpc_url,
    )?;
    let data = CallData {
        contract: context.resolved.contract_name,
        address: context.address,
        function: args.function.clone(),
        signature,
        raw,
    };
    if cli.json {
        let mut meta = Meta::new("call");
        meta.network = Some(context.network);
        meta.account = Some(context.account);
        output::print_json(data, meta)
    } else {
        println!("{} {} -> {}", data.contract, data.signature, data.raw);
        Ok(())
    }
}

pub fn send(cli: &Cli, args: &SendArgs) -> AppResult<()> {
    let context = context(cli, &args.target)?;
    let signature = resolve_function_signature(&context.artifact, &args.function, true)?;
    let gas_estimate = cast_estimate(
        &context.address,
        &signature,
        &args.args,
        args.value.as_deref(),
        &context.network.rpc_url,
    );
    let tx_output = cast_send(
        &context.address,
        &signature,
        &args.args,
        args.value.as_deref(),
        &context.network.rpc_url,
    )?;
    let data = SendData {
        contract: context.resolved.contract_name,
        address: context.address,
        function: args.function.clone(),
        signature,
        tx_output,
        gas_estimate,
    };
    if cli.json {
        let mut meta = Meta::new("send");
        meta.network = Some(context.network);
        meta.account = Some(context.account);
        output::print_json(data, meta)
    } else {
        if let Some(gas) = &data.gas_estimate {
            println!("estimated gas: {gas}");
        }
        println!("{}", data.tx_output);
        Ok(())
    }
}

pub fn state(cli: &Cli, args: &StateArgs) -> AppResult<()> {
    if args.watch {
        return Err(AppError::not_implemented("state --watch"));
    }

    let context = context(cli, &args.target)?;
    let readers = no_arg_readers(&context.artifact);
    let mut values = Vec::new();
    for signature in readers {
        let raw = cast_call(&context.address, &signature, &[], &context.network.rpc_url)?;
        let name = signature
            .split_once('(')
            .map_or(signature.as_str(), |(name, _)| name)
            .to_string();
        values.push(StateValue {
            name,
            signature,
            raw,
        });
    }

    let data = StateData {
        contract: context.resolved.contract_name,
        address: context.address,
        values,
    };
    if cli.json {
        let mut meta = Meta::new("state");
        meta.network = Some(context.network);
        meta.account = Some(context.account);
        output::print_json(data, meta)
    } else {
        println!("{} {}", data.contract, data.address);
        for value in data.values {
            println!("  {:<32} {}", value.name, value.raw);
        }
        Ok(())
    }
}

struct Context {
    resolved: target::ResolvedTarget,
    artifact: Value,
    address: String,
    network: crate::output::NetworkMeta,
    account: crate::output::AccountMeta,
}

fn context(cli: &Cli, target_value: &str) -> AppResult<Context> {
    let resolved = target::resolve(cli, Some(target_value))?;
    let artifact_path = target::artifact_path(&resolved)?;
    let artifact: Value = serde_json::from_str(&fs::read_to_string(artifact_path)?)?;
    let network = detect::active_network(cli)?;
    let account = detect::active_account(cli);
    let deployments = cache::load(&resolved.project_root)?;
    let entry = cache::latest_for_contract(&deployments, &resolved, &network, &account)
        .ok_or_else(|| {
            AppError::user(
                "deployment_not_found",
                format!(
                    "No deployment found for {} on {}.",
                    resolved.contract_name, network.name
                ),
                Some("Run `consol deploy <target>` first.".to_string()),
            )
        })?;
    if !deploy::has_code(&entry.address, &network.rpc_url) {
        return Err(AppError::user(
            "deployment_stale",
            format!(
                "Cached address {} has no code on {}.",
                entry.address, network.name
            ),
            Some("Redeploy the contract for the active network.".to_string()),
        ));
    }
    Ok(Context {
        resolved,
        artifact,
        address: entry.address,
        network,
        account,
    })
}

fn resolve_function_signature(
    artifact: &Value,
    function: &str,
    allow_write: bool,
) -> AppResult<String> {
    if function.contains('(') {
        return Ok(function.to_string());
    }

    let mut matches = Vec::new();
    for item in abi_items(artifact) {
        if item.get("type").and_then(Value::as_str) != Some("function") {
            continue;
        }
        if item.get("name").and_then(Value::as_str) != Some(function) {
            continue;
        }
        let mutability = item
            .get("stateMutability")
            .and_then(Value::as_str)
            .unwrap_or("nonpayable");
        let is_read = mutability == "view" || mutability == "pure";
        if !allow_write && !is_read {
            return Err(AppError::user(
                "function_requires_send",
                format!("Function `{function}` is not view/pure."),
                Some("Use `consol send` for write functions.".to_string()),
            ));
        }
        matches.push(signature(item));
    }

    match matches.as_slice() {
        [signature] => Ok(signature.clone()),
        [] => Err(AppError::user(
            "function_not_found",
            format!("Function `{function}` was not found in the ABI."),
            Some("Run `consol inspect <target>` to list functions.".to_string()),
        )),
        _ => Err(AppError::user(
            "function_ambiguous",
            format!("Function `{function}` is overloaded."),
            Some(format!(
                "Use a full signature. Candidates: {}",
                matches.join(", ")
            )),
        )),
    }
}

fn no_arg_readers(artifact: &Value) -> Vec<String> {
    abi_items(artifact)
        .into_iter()
        .filter(|item| item.get("type").and_then(Value::as_str) == Some("function"))
        .filter(|item| {
            matches!(
                item.get("stateMutability").and_then(Value::as_str),
                Some("view" | "pure")
            )
        })
        .filter(|item| {
            item.get("inputs")
                .and_then(Value::as_array)
                .is_none_or(Vec::is_empty)
        })
        .map(signature)
        .collect()
}

fn abi_items(artifact: &Value) -> Vec<&Value> {
    artifact
        .get("abi")
        .and_then(Value::as_array)
        .map_or_else(Vec::new, |items| items.iter().collect())
}

fn signature(item: &Value) -> String {
    let name = item.get("name").and_then(Value::as_str).unwrap_or_default();
    let inputs = item
        .get("inputs")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|input| input.get("type").and_then(Value::as_str))
        .collect::<Vec<_>>()
        .join(",");
    format!("{name}({inputs})")
}

fn cast_call(address: &str, signature: &str, args: &[String], rpc_url: &str) -> AppResult<String> {
    let mut command = Command::new("cast");
    command
        .arg("call")
        .arg(address)
        .arg(signature)
        .args(args)
        .arg("--rpc-url")
        .arg(rpc_url);
    let output = command.output()?;
    if output.status.success() {
        Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        Err(AppError::user(
            "call_failed",
            format!("cast call failed for {signature}."),
            Some(String::from_utf8_lossy(&output.stderr).to_string()),
        ))
    }
}

fn cast_estimate(
    address: &str,
    signature: &str,
    args: &[String],
    value: Option<&str>,
    rpc_url: &str,
) -> Option<String> {
    let mut command = Command::new("cast");
    command
        .arg("estimate")
        .arg(address)
        .arg(signature)
        .args(args)
        .arg("--rpc-url")
        .arg(rpc_url)
        .arg("--private-key")
        .arg(private_key());
    if let Some(value) = value {
        command.arg("--value").arg(value);
    }
    let output = command.output().ok()?;
    if output.status.success() {
        Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        None
    }
}

fn cast_send(
    address: &str,
    signature: &str,
    args: &[String],
    value: Option<&str>,
    rpc_url: &str,
) -> AppResult<String> {
    let mut command = Command::new("cast");
    command
        .arg("send")
        .arg(address)
        .arg(signature)
        .args(args)
        .arg("--rpc-url")
        .arg(rpc_url)
        .arg("--private-key")
        .arg(private_key());
    if let Some(value) = value {
        command.arg("--value").arg(value);
    }
    let output = command.output()?;
    if output.status.success() {
        Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        Err(AppError::user(
            "send_failed",
            format!("cast send failed for {signature}."),
            Some(String::from_utf8_lossy(&output.stderr).to_string()),
        ))
    }
}

fn private_key() -> String {
    std::env::var("ETH_PRIVATE_KEY").unwrap_or_else(|_| {
        "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80".to_string()
    })
}

File: packages/cli/src/commands/interact.ts
import { parseFunctionItem, ProjectError, readStateKeyBook, type ResolvedTarget } from "@consol/core";
import { runCastCall, runCastDecodeAbi } from "@consol/foundry";
import { createSuccessEnvelope } from "@consol/protocol";
import type { FunctionItem } from "@consol/core";
import { createRpcAdapter } from "@consol/rpc";
import { basename, dirname } from "node:path";
import type { GlobalArgs } from "../args";
import type { CliEnv, CliResult } from "../main";
import { VERSION } from "../version";
import { createReadContext } from "./interact-context";
import { foundryResultMessage, runForgeInspectStorageLayoutWithCacheRecovery } from "./storage-layout-inspect";
import { createComplexStorageSnapshot, type ComplexStorageRow, type ComplexStorageSnapshot } from "./storage-state";

export type CallData = {
  readonly contract: string;
  readonly address: string;
  readonly function: string;
  readonly signature: string;
  readonly raw: string;
};

export type RunCallCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
};

export type StateData = {
  readonly contract: string;
  readonly address: string;
  readonly values: readonly StateValue[];
  readonly storage_values?: readonly ComplexStorageRow[];
  readonly storage_hints?: readonly string[];
  readonly storage_layout_id?: string | null;
};

export type StateValue = {
  readonly name: string;
  readonly signature: string;
  readonly output_types: readonly string[];
  readonly readable: string | null;
  readonly raw: string;
  readonly error?: string | null;
};

export type RunStateCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
  readonly ensureArtifact?: (resolved: ResolvedTarget) => Promise<void>;
};

type CallOptions = {
  readonly target: string;
  readonly functionName: string;
  readonly args: readonly string[];
};

type StateOptions = {
  readonly target: string;
  readonly watch: boolean;
  readonly address?: string;
};

export async function runCallCommand(input: RunCallCommandInput): Promise<CliResult> {
  const options = parseCallOptions(input.commandArgs);
  const context = await createReadContext({
    globals: input.globals,
    cwd: input.cwd,
    env: input.env,
    target: options.target,
  });

  const signature = resolveFunctionSignature(context.artifact.abi, options.functionName);
  const call = await runCastCall({
    cwd: context.resolved.projectRoot,
    env: input.env,
    rpcUrl: context.rpc_url,
    address: context.address,
    signature,
    args: options.args,
  });
  if (!call.ok) {
    return { exitCode: 1, stdout: "", stderr: `cast call failed for ${signature}.\n` };
  }

  const data: CallData = {
    contract: context.resolved.contractName,
    address: context.address,
    function: options.functionName,
    signature,
    raw: call.stdout.trim(),
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "call",
        network: context.network,
        account: context.account,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return { exitCode: 0, stdout: `${data.contract} ${data.signature} -> ${data.raw}\n`, stderr: "" };
}

export async function runStateCommand(input: RunStateCommandInput): Promise<CliResult> {
  const options = parseStateOptions(input.commandArgs);
  if (options.watch && input.globals.json && !input.globals.ndjson) {
    throw new ProjectError({
      code: "ndjson_required",
      message: "`consol state --watch` is a stream.",
      hint: "Use `--ndjson` for watch output, or omit `--watch` for one JSON snapshot.",
    });
  }
  if (options.watch) {
    throw watchNotImplemented("state");
  }

  const context = await createReadContext({
    globals: input.globals,
    cwd: input.cwd,
    env: input.env,
    target: options.target,
    ...(options.address === undefined ? {} : { addressOverride: options.address }),
    ...(input.ensureArtifact === undefined ? {} : { ensureArtifact: input.ensureArtifact }),
  });
  const values: StateValue[] = [];
  for (const reader of noArgumentReaders(context.artifact.abi)) {
    const call = await runCastCall({
      cwd: context.resolved.projectRoot,
      env: input.env,
      rpcUrl: context.rpc_url,
      address: context.address,
      signature: reader.signature,
      args: [],
    });
    if (!call.ok) {
      values.push({
        name: reader.name,
        signature: reader.signature,
        output_types: reader.outputs.map((output) => output.kind),
        readable: null,
        raw: "",
        error: castCallFailureMessage(reader.signature),
      });
      continue;
    }

    const raw = call.stdout.trim();
    values.push({
      name: reader.name,
      signature: reader.signature,
      output_types: reader.outputs.map((output) => output.kind),
      readable: await decodeReadable({
        cwd: context.resolved.projectRoot,
        env: input.env,
        outputTypes: reader.outputs.map((output) => output.kind),
        raw,
      }),
      raw,
    });
  }

  const complex = filterComplexStorageState(await maybeCreateComplexStorageState({ context, input }), values);
  const data: StateData = {
    contract: context.resolved.contractName,
    address: context.address,
    values,
    ...(complex === null
      ? {}
      : {
        storage_values: complex.rows,
        storage_hints: complex.hints,
        storage_layout_id: complex.layout_id,
      }),
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "state",
        network: context.network,
        account: context.account,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return { exitCode: 0, stdout: stateHuman(data), stderr: "" };
}

function stateHuman(data: StateData): string {
  const lines = [`${data.contract} ${data.address}`];
  if (data.values.length === 0) {
    lines.push("  (no no-argument readers)");
    return `${lines.join("\n")}\n`;
  }

  for (const value of data.values) {
    const displayValue = value.error === undefined || value.error === null ? value.readable ?? value.raw : `! ${value.error}`;
    lines.push(`  ${value.signature} = ${displayValue}`);
  }
  return `${lines.join("\n")}\n`;
}

async function maybeCreateComplexStorageState(input: {
  readonly context: Awaited<ReturnType<typeof createReadContext>>;
  readonly input: RunStateCommandInput;
}): Promise<Awaited<ReturnType<typeof createComplexStorageSnapshot>> | null> {
  try {
    const contractId = contractIdentifier(
      input.context.artifact.raw,
      input.context.artifact.path,
      input.context.resolved.contractName,
    );
    const layout = await runForgeInspectStorageLayoutWithCacheRecovery({
      cwd: input.context.resolved.projectRoot,
      projectRoot: input.context.resolved.projectRoot,
      contractId,
      env: input.input.env,
    });
    if (!layout.ok) {
      return storageLayoutFailureSnapshot(foundryResultMessage(layout));
    }

    return await createComplexStorageSnapshot({
      layoutJson: layout.stdout,
      projectRoot: input.context.resolved.projectRoot,
      target: stateTarget(input.context.resolved),
      contract: input.context.resolved.contractName,
      address: input.context.address,
      rpc: createRpcAdapter({ rpcUrl: input.context.rpc_url }),
      keyBook: readStateKeyBook(input.context.resolved.projectRoot),
      previewLimit: 3,
      mode: "summary",
    });
  } catch (error) {
    return storageLayoutFailureSnapshot(error instanceof Error ? error.message : String(error));
  }
}

function storageLayoutFailureSnapshot(message: string): ComplexStorageSnapshot {
  const summary = message.trim().length === 0 ? "Storage layout is unavailable." : message.trim();
  return {
    layout_id: "layout:error",
    rows: [{
      id: "storage:layout:error",
      kind: "error",
      name: "storage layout",
      type_label: "storage-layout",
      summary,
      detail_available: false,
      error: summary,
    }],
    hints: ["storage layout unavailable; run forge build and refresh"],
  };
}

function filterComplexStorageState(
  snapshot: ComplexStorageSnapshot | null,
  values: readonly StateValue[],
): ComplexStorageSnapshot | null {
  if (snapshot === null) {
    return null;
  }

  const abiReaderNames = new Set(values.map((value) => value.name));
  const rows = snapshot.rows.filter((row) => row.kind !== "scalar" || !abiReaderNames.has(row.name));
  return rows.length === 0 && snapshot.hints.length === 0 ? null : { ...snapshot, rows };
}

function stateTarget(resolved: ResolvedTarget): string {
  return resolved.sourceFile === undefined ? resolved.contractName : `${resolved.sourceFile}:${resolved.contractName}`;
}

function contractIdentifier(rawArtifact: unknown, artifactPath: string, contractName: string): string {
  const source = artifactSource(rawArtifact);
  if (source !== undefined) {
    return `${source}:${contractName}`;
  }

  return `src/${basename(dirname(artifactPath))}:${contractName}`;
}

function artifactSource(rawArtifact: unknown): string | undefined {
  const metadata = getRecordProperty(rawArtifact, "metadata");
  const settings = getRecordProperty(metadata, "settings");
  const compilationTarget = getRecordProperty(settings, "compilationTarget");
  return compilationTarget === undefined ? undefined : Object.keys(compilationTarget)[0];
}

function castCallFailureMessage(signature: string): string {
  return `cast call failed for ${signature}.`;
}

function parseCallOptions(commandArgs: readonly string[]): CallOptions {
  const args = commandArgs.filter((arg) => arg !== "--json");
  const target = args[0];
  const functionName = args[1];
  if (target === undefined || functionName === undefined) {
    throw new ProjectError({
      code: "call_args_required",
      message: "Missing target or function for call.",
      hint: "Use `consol call <target> <viewFunction> [args...]`.",
    });
  }

  return {
    target,
    functionName,
    args: args.slice(2),
  };
}

function parseStateOptions(commandArgs: readonly string[]): StateOptions {
  let target: string | undefined;
  let address: string | undefined;
  let watch = false;

  for (let index = 0; index < commandArgs.length; index += 1) {
    const arg = commandArgs[index];
    if (arg === undefined || arg === "--json" || arg === "--ndjson") {
      continue;
    }
    if (arg === "--watch") {
      watch = true;
      continue;
    }
    if (arg === "--address") {
      const nextAddress = commandArgs[index + 1];
      if (nextAddress === undefined) {
        throw new ProjectError({
          code: "missing_flag_value",
          message: "Missing value for --address.",
          hint: "Pass a deployed contract address after --address.",
        });
      }
      address = nextAddress;
      index += 1;
      continue;
    }
    if (target === undefined) {
      target = arg;
    }
  }

  if (target === undefined) {
    throw new ProjectError({
      code: "state_target_required",
      message: "Missing target for state.",
      hint: "Use `consol state <target>`.",
    });
  }

  return {
    target,
    watch,
    ...(address === undefined ? {} : { address }),
  };
}

function watchNotImplemented(command: "state"): ProjectError {
  return new ProjectError({
    code: "watch_not_implemented",
    message: `\`consol ${command} --watch\` needs a streaming runner before it can be used safely.`,
    hint: `Omit \`--watch\` for a one-shot snapshot until ${command} streaming is implemented.`,
  });
}

export function resolveFunctionSignature(abi: readonly unknown[], functionName: string, allowWrite = false): string {
  if (functionName.includes("(")) {
    return functionName;
  }

  const matches = abi
    .filter((item) => getStringProperty(item, "type") === "function")
    .filter((item) => getStringProperty(item, "name") === functionName)
    .map(parseFunctionItem);
  const candidates = allowWrite ? matches : matches.filter(isReadableFunction);
  if (matches.length > 0 && candidates.length === 0) {
    throw new ProjectError({
      code: "function_requires_send",
      message: `Function \`${functionName}\` is not view/pure.`,
      hint: "Use `consol send` for write functions.",
    });
  }

  if (candidates.length === 1) {
    return candidates[0]?.signature ?? unreachable("expected one function signature");
  }

  if (candidates.length === 0) {
    throw new ProjectError({
      code: "function_not_found",
      message: `Function \`${functionName}\` was not found in the ABI.`,
      hint: "Run `consol inspect <target>` to list functions.",
    });
  }

  throw new ProjectError({
    code: "function_ambiguous",
    message: `Function \`${functionName}\` is overloaded.`,
    hint: `Use a full signature. Candidates: ${candidates.map((item) => item.signature).join(", ")}`,
  });
}

function isReadableFunction(item: FunctionItem): boolean {
  return item.state_mutability === "view" || item.state_mutability === "pure";
}

function noArgumentReaders(abi: readonly unknown[]): readonly FunctionItem[] {
  return abi
    .filter((item) => getStringProperty(item, "type") === "function")
    .map(parseFunctionItem)
    .filter(isReadableFunction)
    .filter((item) => item.inputs.length === 0);
}

async function decodeReadable(input: {
  readonly cwd: string;
  readonly env: CliEnv;
  readonly outputTypes: readonly string[];
  readonly raw: string;
}): Promise<string | null> {
  if (input.outputTypes.length === 0) {
    return null;
  }

  const result = await runCastDecodeAbi({
    cwd: input.cwd,
    env: input.env,
    signature: `__consol_decode()(${input.outputTypes.join(",")})`,
    data: input.raw,
  });
  if (!result.ok) {
    return null;
  }

  const values = result.stdout
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
  return values.length === 0 ? null : values.join(", ");
}

function getStringProperty(raw: unknown, key: string): string | undefined {
  const value = getProperty(raw, key);
  return typeof value === "string" ? value : undefined;
}

function getRecordProperty(raw: unknown, key: string): Record<string, unknown> | undefined {
  const value = getProperty(raw, key);
  return typeof value === "object" && value !== null && !Array.isArray(value) ? value as Record<string, unknown> : undefined;
}

function getProperty(raw: unknown, key: string): unknown {
  return typeof raw === "object" && raw !== null && !Array.isArray(raw) ? (raw as Record<string, unknown>)[key] : undefined;
}

function unreachable(message: string): never {
  throw new Error(message);
}
